main.go

```diff
@@ -186,87 +186,71 @@ func getSecretGenerators(c *controller) []SecretGenerator {
 	return secretGenerators
 }
 
-func (c *controller) process() error {
-	secretGenerators := getSecretGenerators(c)
-
-	for _, secretGenerator := range secretGenerators {
+func (c *controller) processNamespace(namespace *v1.Namespace, secret *v1.Secret) error {
+	// Check if the secret exists for the namespace
+	_, err := c.k8sutil.GetSecret(namespace.GetName(), secret.Name)
 
-		fmt.Printf("------------------ [%s] ----------------------\n", secretGenerator.SecretName)
-
-		newToken, err := secretGenerator.TokenGenFxn()
+	if err != nil {
+		// Secret not found, create
+		err := c.k8sutil.CreateSecret(namespace.GetName(), secret)
 		if err != nil {
-			fmt.Printf("Error getting secret for provider %s. Skipping secret provider! [Err: %s]", secretGenerator.SecretName, err)
-			continue
+			return fmt.Errorf("Could not create Secret! %v", err)
 		}
-		newSecret := generateSecretObj(newToken.AccessToken, newToken.Endpoint, secretGenerator.IsJSONCfg, secretGenerator.SecretName)
-
-		namespaces, err := c.k8sutil.GetNamespaces()
-
+	} else {
+		// Existing secret needs updated
+		err := c.k8sutil.UpdateSecret(namespace.GetName(), secret)
 		if err != nil {
-			fmt.Println("-------> ERROR getting namespaces! Skipping secret provider!", err)
-			continue
+			return fmt.Errorf("Could not update Secret! %v", err)
 		}
+	}
+
+	// Check if ServiceAccount exists
+	serviceAccount, err := c.k8sutil.GetServiceAccount(namespace.GetName(), "default")
+	if err != nil {
+		return fmt.Errorf("Could not get ServiceAccounts! %v", err)
```

Comment:
I think the behavior is unchanged here.  In the [logic in master](https://github.com/upmc-enterprises/registry-creds/blob/master/main.go#L236-L241), the namespaces are processed in a loop and the `continue` moves onto the next *namespace*.  Since we are only processing namespaces one at a time as they are added or modified (due to the watch), I think returning on error is probably appropriate.  This error [gets bubbled up](https://github.com/upmc-enterprises/registry-creds/pull/40/files?diff=split#diff-7ddfb3e035b42cd70649cc33393fe32cR277) to the `handler()` function below, which iterates over all the secrets.  I think everything in `processNamespace()` is *namespace specific*, though so trying again with the next secret isn't going to change the outcome.

If I've misunderstood/read something, I'm happy to change it!